nav2_collision_monitor/src/source_base.cpp

```diff
@@ -0,0 +1,192 @@
+    source_topic_ = node->get_parameter(source_name_ + ".topic").as_string();
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      node->get_logger(), "Error while getting basic source parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+SourceType SourceBase::getSourceType()
+{
+  return source_type_;
+}
+
+std::string SourceBase::getSourceTypeStr()
+{
+  if (source_type_ == SOURCE_BASE) {
+    return "base";
+  } else if (source_type_ == SCAN) {
+    return "scan";
+  } else {  // source_type_ == POINTCLOUD
+    return "pointcloud";
+  }
+}
+
+bool SourceBase::getTransform(
+  const std::string to_frame,
+  const std::string from_frame,
+  tf2::Transform & tf2_transform)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!tf_buffer_) {
+    RCLCPP_ERROR(node->get_logger(), "tf_buffer_ is not initialized");
+    return false;
+  }
+
+  geometry_msgs::msg::TransformStamped transform;
+  tf2_transform.setIdentity();  // initialize by identical transform
+
+  if (from_frame == to_frame) {
+    // tf2_transform is already identical. Do nothing.
+    return true;
+  }
+
+  try {
+    const tf2::Duration tr_tol = tf2::durationFromSec(transform_tolerance_);
+    // Obtaining the transform to get data from one to another frame
+    transform = tf_buffer_->lookupTransform(
+      to_frame, from_frame, tf2::TimePointZero, tr_tol);
+  } catch (tf2::TransformException & e) {
+    RCLCPP_ERROR(
+      node->get_logger(),
+      "Failed to get \"%s\"->\"%s\" frame transform: %s",
+      source_frame_id_.c_str(), base_frame_id_.c_str(), e.what());
+    return false;
+  }
+
+  // Convert TransformStamped to TF2 transform
+  tf2::fromMsg(transform.transform, tf2_transform);
+  return true;
+}
+
+void SourceBase::getData(
+  std::vector<Point> & data, const rclcpp::Time & curr_time, const Velocity & velocity)
+{
+  std::lock_guard<mutex_t> lock(data_mutex_);
+
+  fixData(curr_time, velocity);
+  data.insert(data.end(), data_fixed_.data.begin(), data_fixed_.data.end());
+}
+
+void SourceBase::fixData(const rclcpp::Time & curr_time, const Velocity & velocity)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // Start opearting with data
+  std::lock_guard<mutex_t> lock(data_mutex_);
+  data_fixed_.data.clear();
+  data_fixed_.time = data_.time;
+
+  // Time shift between current time and time of latest data received.
+  // Used in movement correction of data
+  double dt;
+  try {
+    dt = (curr_time - data_.time).seconds();
+  } catch (const std::exception & ex) {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "Can not get difference between collision monitor and %s source time: %s",
+      source_name_.c_str(), ex.what());
+    dt = 0.0;
+  }
+  // Warn and ignore the source data if it and Collision Monitor node have different time stamps
+  if (std::fabs(dt) > max_time_shift_) {
```

Comment:
This can all be done with ROS types `curr_time - data_.time > rclcpp::Duration` if `max_time_shift_` is stored as a duration type, significantly shortening this function

Reply:
Done in latest update